doc/man/bitcoin-cli.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH BITCOIN-CLI "1" "March 2026" "bitcoin-cli v1.2.1.0" "User Commands"
.SH NAME
bitcoin-cli \- manual page for bitcoin-cli v1.2.1.0
.SH DESCRIPTION
Bitcoin SV RPC client version v1.2.1.0\-dev
.SS "Usage:"
.IP
bitcoin\-cli [options] Send command to Bitcoin SV
bitcoin\-cli [options] help                List commands
bitcoin\-cli [options] help <command>      Get help for a command
.SH OPTIONS
.HP
\-?
.IP
This help message
.HP
\fB\-conf=\fR<file>
.IP
Specify configuration file (default: bitcoin.conf)
.HP
\fB\-datadir=\fR<dir>
.IP
Specify data directory
.PP
Chain selection options:
.HP
\fB\-testnet\fR
.IP
Use the test chain
.HP
\fB\-regtest\fR
.IP
Enter regression test mode, which uses a special chain in which blocks
can be solved instantly. This is intended for regression testing
tools and app development.
.HP
\fB\-stn\fR
.IP
Use the Scaling Test Network
.PP
RPC Options:
.HP
\fB\-named\fR
.IP
Pass named parameters instead of positional arguments (default: false)
.HP
\fB\-rpcconnect=\fR<ip>
.IP
Send commands to node running on <ip> (default: 127.0.0.1)
.HP
\fB\-rpcport=\fR<port>
.IP
Connect to JSON\-RPC on <port> (default: 8332 or testnet: 18332)
.HP
\fB\-rpcwait\fR
.IP
Wait for RPC server to start
.HP
\fB\-rpcuser=\fR<user>
.IP
Username for JSON\-RPC connections
.HP
\fB\-rpcpassword=\fR<pw>
.IP
Password for JSON\-RPC connections
.HP
\fB\-rpcclienttimeout=\fR<n>
.IP
Timeout in seconds during HTTP requests, or 0 for no timeout. (default:
900)
.HP
\fB\-stdinrpcpass\fR
.TP
Read RPC password from standard input as a single line.
When combined
.IP
with \fB\-stdin\fR, the first line from standard input is used for the
RPC password.
.HP
\fB\-stdin\fR
.IP
Read extra arguments from standard input, one per line until EOF/Ctrl\-D
(recommended for sensitive information such as passphrases)
.HP
\fB\-rpcwallet=\fR<walletname>
.IP
Send RPC for non\-default wallet on RPC server (argument is wallet
filename in bitcoind directory, required if bitcoind runs with
multiple wallets)
.SH COPYRIGHT
Copyright (C) 2009-2026 The Bitcoin Association

Please contribute if you find Bitcoin SV useful. Visit <https://bitcoinsv.io>
for further information about the software.
The source code is available from <https://github.com/bitcoin-sv/bitcoin-sv>.

This is experimental software.
Distributed under the Open BSV software license, see the accompanying file
LICENSE

This product includes software developed by the OpenSSL Project for use in the
OpenSSL Toolkit <https://www.openssl.org> and cryptographic software written by
Eric Young and UPnP software written by Thomas Bernard.


doc/release-notes.md
# Bitcoin SV Node Software – v1.2.0 Release

This release is a hard fork which includes changes to the BSV protocol.   

The scheduled TestNet activation height is 1,686,611 (Target is 12:00 midday 05-Aug-2025)

The scheduled MainNet activation height is 921,788 (Target is 12:00 midday 04-Nov-2025)

## What’s Changed  

### Chronicle Protocol Updates

#### Script OpCodes

The following opcodes are restored:  

*   **OP\_VER** – pushes the transaction version (4 byte array) onto the top of the stack.
*   **OP\_VERIF, OP\_VERNOTIF** – provides conditional logic based on the transaction version.
*   **OP\_SUBSTR** – creates an arbitrary substring of the string on top of the stack.
*   **OP\_LEFT, OP\_RIGHT –** creates the left/right most substring of specific length from the string on the top of the stack respectively.
*   **OP\_2MUL** – doubles the number on top of the stack
*   **OP\_2DIV** – halves the number on top of the stack

Please refer to the Protocol Restoration specification for further details.  

#### Sighash Flag Changes

A new optional sighash flag **SIGHASH_CHRONICLE** is introduced (value **0x20**).

Transactions signed without the CHRONICLE flag will see no changes to their
consensus rules and will remain largely unaffected by the Chronicle update.
Transactions signed *with* the CHRONICLE flag will see the changes described
in the following sections.

#### Transaction Digest Algorithm

The BSV blockchain will again support the original Bitcoin transaction digest
algorithm for transactions signed with the CHRONICLE sighash flag. Transactions
signed without the CHRONICLE flag will continue to use the BIP143 new transaction
digest algorithm as before.

#### Transaction Limitations Relaxed

The limitations in the following areas are removed.   

*   **Script numbers** – Limits on the maximum size of script numbers are removed.
Please note that practical limits are imposed by the external libraries used to
implement script numbers (max size is currently 64MB).
*   **Minimal Encoding Requirement** – Previously releases required transactions
to encode numbers as efficiently as possible. For example, the number two can be
encoded as 0002 (2 bytes) or 02 (1 byte). Prior to this release only the second
version is accepted. Minimal encoding places an unnecessary burden on users that
was not present in the original Satoshi implementation and it is removed from this
release for transactions signed with the CHRONICLE sighash flag.
*   **Low S signatures** - If S is a signature, then so is -S. The “low S” signature
requirement was introduced to decrease transaction malleability but is an unnecessary
requirement and is removed. Either signature S or -S can now be used in transactions
signed with the CHRONICLE sighash flag.
*   **NULLFAIL checks** - The requirement that if an `OP_CHECKSIG` or
`OP_CHECKMULTISIG` is trying to return a `FALSE` value to the stack then signatures
must be an empty array is removed for transactions signed with the CHRONICLE sighash
flag.
*   **NULLDUMMY checks** - The requirement that the dummy value in the unlocking
script for a multisig transaction must be an empty array is removed for transactions
signed with the CHRONICLE sighash flag.
*   **Clean Stack Policy** – Previous releases required that after the execution
of the unlocking and locking script, the stack is “clean”. i.e. there is only a
single item (interpreted as true/false) on the stack that indicates whether the
transaction has permission to spend the relevant input. The requirement is
unnecessary and adds complexity to scripts. This release removes the clean stack
requirement for transactions signed with the CHRONICLE sighash flag.
*   **No opcodes in Unlocking Scripts Policy** - Previous releases do not allow
unlocking scripts to include non-data opcodes. That restriction is removed in
this release for transactions signed with the CHRONICLE sighash flag.

### Performance Improvements

#### Memory Usage

TCMalloc is now the default memory allocator used when building bitcoind. In
testing, this change results in peak memory usage of the node dropping by nearly
a half.

#### Initial Block Download (IBD)

By tuning some existing leveldb configuration parameters, introducing some new
configuration options and optimising the code, the time taken to perform a
complete IBD has been reduced. See the section below for details on the new
config options.

### Config Variables

#### Mempool Synchonisation

A new feature has been introduced to allow nodes to be configured to periodically
update their mempools with missing transactions from other specified peers. Note
that peers that wish to synchronise mempools must be configured symmetrically, i.e.
both peers must be configured to sync' with each other.

The options to configure this are:

*   **mempoolsyncpeer** - Specify IPs or subnets of peers with which we want to
periodically synchronise mempools. Can be used more than once to specify
multiple peers / networks.
*   **mempoolsyncage** - Only transactions older than this age will be included
in synchronisation.
*   **mempoolsyncperiod** - How often synchronisation takes place.

#### Other New Options

*   **preferredblockfilesize** - Preferred size (in bytes) of a single block datafile.
*   **maxcoinsdbfilesize** - Set maximum file size used by the coins database.
*   **maxconnections** - Maintain at most the given number of connections to peers.
*   **maxchroniclegracefulperiod** - The number of blocks either side of the Chronicle
activation height that we will give peers some grace if they are not yet fully
activated before banning them.

#### Removed Options

*   **maxsendbuffermult**

#### Changes to Existing Defaults

*   **dbcache** - The default database cache size is now 2GB.
*   **maxscriptnumlengthpolicy** - The default is now 0 (unlimited).

### Other items

*   STN Reset - includes an updated chain height block hash.

src/bitcoin-cli.cpp
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2016 The Bitcoin Core developers
// Copyright (c) 2019 Bitcoin Association
// Distributed under the Open BSV software license, see the accompanying file LICENSE.

#if defined(HAVE_CONFIG_H)
#include "config/bitcoin-config.h"
#endif

#include "chainparamsbase.h"
#include "clientversion.h"
#include "rpc/client_utils.h"
#include "rpc/protocol.h"
#include "util.h"
#include <cstdio>
#include <univalue.h>

std::string HelpMessageCli() {
    const auto defaultBaseParams =
        CreateBaseChainParams(CBaseChainParams::MAIN);
    const auto testnetBaseParams =
        CreateBaseChainParams(CBaseChainParams::TESTNET);
    std::string strUsage;
    strUsage += HelpMessageGroup(_("Options:"));
    strUsage += HelpMessageOpt("-?", _("This help message"));
    strUsage += HelpMessageOpt(
        "-conf=<file>", strprintf(_("Specify configuration file (default: %s)"),
                                  BITCOIN_CONF_FILENAME));
    strUsage += HelpMessageOpt("-datadir=<dir>", _("Specify data directory"));
    AppendParamsHelpMessages(strUsage);
    strUsage += HelpMessageGroup(_("RPC Options:"));
    strUsage += HelpMessageOpt(
        "-named",
        strprintf(_("Pass named parameters instead of positional arguments (default: %s)"),
                  DEFAULT_NAMED));
    strUsage += HelpMessageOpt(
        "-rpcconnect=<ip>",
        strprintf(_("Send commands to node running on <ip> (default: %s)"),
                  DEFAULT_RPCCONNECT));
    strUsage += HelpMessageOpt(
        "-rpcport=<port>",
        strprintf(
            _("Connect to JSON-RPC on <port> (default: %u or testnet: %u)"),
            defaultBaseParams->RPCPort(), testnetBaseParams->RPCPort()));
    strUsage += HelpMessageOpt("-rpcwait", _("Wait for RPC server to start"));
    strUsage += HelpMessageOpt("-rpcuser=<user>",
                               _("Username for JSON-RPC connections"));
    strUsage += HelpMessageOpt("-rpcpassword=<pw>",
                               _("Password for JSON-RPC connections"));
    strUsage +=
        HelpMessageOpt("-rpcclienttimeout=<n>",
                       strprintf(_("Timeout in seconds during HTTP requests, "
                                   "or 0 for no timeout. (default: %d)"),
                                 DEFAULT_HTTP_CLIENT_TIMEOUT));

    strUsage += HelpMessageOpt(
        "-stdinrpcpass",
        strprintf(_("Read RPC password from standard input as a single line.  "
                    "When combined with -stdin, the first line from standard "
                    "input is used for the RPC password.")));
    strUsage += HelpMessageOpt(
        "-stdin", _("Read extra arguments from standard input, one per line "
                    "until EOF/Ctrl-D (recommended for sensitive information "
                    "such as passphrases)"));
    strUsage += HelpMessageOpt(
        "-rpcwallet=<walletname>",
        _("Send RPC for non-default wallet on RPC server (argument is wallet "
          "filename in bitcoind directory, required if bitcoind runs with "
          "multiple wallets)"));

    return strUsage;
}

//////////////////////////////////////////////////////////////////////////////
//
// Start
//

// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
int CommandLineRPC(int argc, char *argv[]) {
    std::string strPrint;
    int nRet = 0;
    try {
        // Skip switches
        while (argc > 1 && IsSwitchChar(argv[1][0])) { // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
            argc--;
            argv++; // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
        }
        std::string rpcPass;
        if (gArgs.GetBoolArg("-stdinrpcpass", false)) {
            if (!std::getline(std::cin, rpcPass))
                throw std::runtime_error("-stdinrpcpass specified but failed "
                                         "to read from standard input");
            gArgs.ForceSetArg("-rpcpassword", rpcPass);
        }
        std::vector<std::string> args =
            std::vector<std::string>(&argv[1], &argv[argc]); // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
        if (gArgs.GetBoolArg("-stdin", false)) {
            // Read one arg per line from stdin and append
            std::string line;
            while (std::getline(std::cin, line)) {
                args.push_back(line);
            }
        }
        if (args.size() < 1) {
            throw std::runtime_error(
                "too few parameters (need at least command)");
        }
        std::string strMethod = args[0];
        // Remove trailing method name from arguments vector
        args.erase(args.begin());

        UniValue params;
        // NOLINTNEXTLINE(bugprone-branch-clone)
        if (gArgs.GetBoolArg("-named", DEFAULT_NAMED)) {
            params = RPCConvertNamedValues(strMethod, args);
        } else {
            params = RPCConvertValues(strMethod, args);
        }

        // Execute and handle connection failures with -rpcwait
        const bool fWait = gArgs.GetBoolArg("-rpcwait", false);
        do {
            try {
                const UniValue reply = CallRPC(strMethod, params);

                // Parse reply
                const UniValue result = find_value(reply, "result");
                const UniValue error = find_value(reply, "error");

                if (!error.isNull()) {
                    // Error
                    int code = error["code"].get_int();
                    if (fWait && code == RPC_IN_WARMUP)
                        throw CConnectionFailed("server in warmup");
                    strPrint = "error: " + error.write();
                    nRet = abs(code);
                    if (error.isObject()) {
                        UniValue errCode = find_value(error, "code");
                        UniValue errMsg = find_value(error, "message");
                        strPrint =
                            errCode.isNull()
                                ? ""
                                : "error code: " + errCode.getValStr() + "\n";

                        if (errMsg.isStr()) {
                            strPrint += "error message:\n" + errMsg.get_str();
                        }

                        if (errCode.isNum() &&
                            errCode.get_int() == RPC_WALLET_NOT_SPECIFIED) {
                            strPrint += "\nTry adding "
                                        "\"-rpcwallet=<filename>\" option to "
                                        "bitcoin-cli command line.";
                        }
                    }
                } else {
                    // Result
                    if (result.isNull()) {
                        strPrint = "";
                    } else if (result.isStr()) {
                        strPrint = result.get_str();
                    } else {
                        strPrint = result.write(2);
                    }
                }
                // Connection succeeded, no need to retry.
                break;
            } catch (const CConnectionFailed &) {
                if (fWait) {
                    MilliSleep(1000);
                } else {
                    throw;
                }
            }
        } while (fWait);
    } catch (const boost::thread_interrupted &) {
        throw;
    } catch (const std::exception &e) {
        strPrint = std::string("error: ") + e.what();
        nRet = EXIT_FAILURE;
    } catch (...) {
        PrintExceptionContinue(nullptr, "CommandLineRPC()");
        throw;
    }

    if(strPrint != "")
    {
        std::ostream& os = (nRet == 0 ? std::cout : std::cerr);
        os << strPrint << '\n';
    }
    return nRet;
}

int main(int argc, char *argv[])
{
    SetupEnvironment();
    if(!SetupNetworking())
    {
        std::cerr << "Error: Initializing networking failed\n";
        return EXIT_FAILURE;
    }

    try 
    {
       std::string appname("bitcoin-cli");
       std::string usage = "\n" + _("Usage:") + "\n" + "  " + appname + " [options] " +
                            strprintf(_("Send command to %s"), _(PACKAGE_NAME)) + "\n" + "  " + appname +
                            " [options] help                " + _("List commands") + "\n" + "  " + appname +
                            " [options] help <command>      " + _("Get help for a command") + "\n";

        int ret = AppInitRPC(argc, argv, usage, HelpMessageCli);
        if (ret != CONTINUE_EXECUTION) 
        {
            return ret;
        }
    } catch (const std::exception &e) {
        PrintExceptionContinue(&e, "AppInitRPC()");
        return EXIT_FAILURE;
    } catch (...) {
        PrintExceptionContinue(nullptr, "AppInitRPC()");
        return EXIT_FAILURE;
    }

    int ret = EXIT_FAILURE;
    try {
        ret = CommandLineRPC(argc, argv);
    } catch (const std::exception &e) {
        PrintExceptionContinue(&e, "CommandLineRPC()");
    } catch (...) {
        PrintExceptionContinue(nullptr, "CommandLineRPC()");
    }
    return ret;
}
